Fix `defaultValues` getting overwritten

rplopes opened the pull request:
New sequelize instances don't get their `defaultValue`s cloned. This means that if we have a model `User` with an attribute `business` with default value `{ status: 'none' }`, then new sequelize instances of `User` will share the same object in `business`, and one change in one instance will also affect all other instances.

Example:

```
const user1 = db.User.build();
user1.business.status; // 'none'
user1.business.status = 'ok';
user1.business.status; // 'ok'

const user2 = db.User.build();
user2.business.status; // 'ok'
```

In this example, `user2.business.status` should clearly be `none` (the default value), but is `ok` instead, because of the change made in `user1`.

This behaviour seems wrong. A developer's expectation of changing `user1.business.status` is that it's only affecting `user1`'s state, and not some global mutable state that is shared among all `User` instances. This could result in unexpected bugs and in wrong tests.

The suggested fix is to clone the default values upon new instance creation (like sequelize is already doing for values passed into the instance constructor).

Commit:
Fix defaultValues getting overwritten

ruiquelhas commented:
Is `_.clone` enough or should we use `_.cloneDeep` instead? What kind of information does the function return? Also, we should add one or more tests for this so we can get it merged easily in the main repo.

rplopes commented:
I took a look at the test suite (in hindsight, I don't know how I didn't do that sooner). I'm stuck on `test/integration/model/create.test.js:781`:

```
  1) [SQLITE] Model create is possible to use functions as default values:
     SequelizeValidationError: string violation: year cannot be an array or an object
      at lib/instance-validator.js:73:14
  From previous event:
      at Promise.then (lib/promise.js:21:17)
      at Context.<anonymous> (test/integration/model/create.test.js:811:53)
```

This test is passing an SQLite function as a default value of the instance, and that isn't working very well with both `_.clone` and `_.cloneDeep`. That is because at that point in time, `valueFn` is:

```
{ fn: '', args: [ { fn: 'date', args: [Object] } ] }
```

(i.e. the SQLite function, still in JSON object format). So by cloning that object, we seem to somehow lose something important, because the value of `year` (the attribute with that default value) will be `undefined`, thus causing that validation error.

Any ideas on why cloning this SQLite function JSON object instead of using it directly makes `year` become `undefined` instead of a string? Is it something important that would make the change in this PR not suitable to be merged?

rplopes commented:
I've updated these changes to make the tests pass, and added a new unit test with what I consider a sensible behaviour, that would fail under the old code.

In response to the question in my previous comment, that integration test was using a sequelize method as a default value. So `valueFn()._isSequelizeMethod` would return `true`. However, `_isSequelizeMethod` doesn't get cloned on `_.clone()` or `_.cloneDeep()`, so `_.cloneDeep(valueFn())._isInstanceMethod` is `undefined`. Later, in validations, because `year` was meant to be a string, its value `{ fn: '', args: [ { fn: 'date', args: [Object] } ] }` was being parsed as an object (this explains the error message: `SequelizeValidationError: string violation: year cannot be an array or an object`). This validation doesn't apply to sequelize methods, hence the difference in behaviour.

In summary, the issue is fixed and I'm again submitting this to be merged.

ruiquelhas commented:
Have you tried using `_.mapValues()` without cloning?

rplopes commented:
@ruiquelhas Yes, and it behaves just like `_.each()` was behaving. Which is ok, I guess, because apparently we don't want to clone values when they're sequelize methods (their `_isSequelizeMethod` attribute isn't cloned, which results in errors later).

nunofgs commented on test/unit/instance/build.test.js:
```diff
@@ -71,5 +71,21 @@ describe(Support.getTestDialectTeaser('Instance'), function() {
+          type: DataTypes.JSONB,
+          defaultValue: { key: 'default value' }
+        }
+      })
```
Missing semicolon.

rplopes replied:
Here, I'm following the convention of this file (see e.g. test above). I believe it's better to follow that convention, specially if we plan to submit this upstream later.

nunofgs commented on test/unit/instance/build.test.js:
```diff
@@ -71,5 +71,21 @@ describe(Support.getTestDialectTeaser('Instance'), function() {
+
+    it('should change values for individual instances only', function () {
+      var Model = current.define('Model', {
+        model_attribute: {
```
I suggest using `data` as the field name.

nunofgs commented on test/unit/instance/build.test.js:
```diff
@@ -71,5 +71,21 @@ describe(Support.getTestDialectTeaser('Instance'), function() {
+      var Model = current.define('Model', {
+        model_attribute: {
+          type: DataTypes.JSONB,
+          defaultValue: { key: 'default value' }
```
Suggestion for defaultValue: `{ foo: 'bar' }`

nunofgs commented on test/unit/instance/build.test.js:
```diff
@@ -71,5 +71,21 @@ describe(Support.getTestDialectTeaser('Instance'), function() {
+          defaultValue: { key: 'default value' }
+        }
+      })
+        , instance;
```
Use `var`.

rplopes replied:
Same as above.

nunofgs commented on test/unit/instance/build.test.js:
```diff
@@ -71,5 +71,21 @@ describe(Support.getTestDialectTeaser('Instance'), function() {
+      })
+        , instance;
+
+      instance = Model.build();
```
This can be moved to the assignment above.

rplopes replied:
Here, I'm following the convention of this file (see e.g. test above). I believe it's better to follow that convention, specially if we plan to submit this upstream later.

nunofgs commented on test/unit/instance/build.test.js:
```diff
@@ -71,5 +71,21 @@ describe(Support.getTestDialectTeaser('Instance'), function() {
+        , instance;
+
+      instance = Model.build();
+      instance.model_attribute.key = 'new value';
```
Suggestion for value: `biz`

nunofgs commented on test/unit/instance/build.test.js:
```diff
@@ -71,5 +71,21 @@ describe(Support.getTestDialectTeaser('Instance'), function() {
       expect(instance.get('id')).to.be.ok;
     });
+
+    it('should change values for individual instances only', function () {
```
Suggestion for test description: `it('should clone the default values')` (or something along those lines).

nunofgs commented on lib/instance.js:
```diff
@@ -22,10 +22,9 @@ var initValues = function(values, options) {
-        }
+      defaults = _.mapValues(this.Model._defaultValues, function(valueFn) {
+        var value = valueFn();
+        return value._isSequelizeMethod ? value : _.cloneDeep(value);
```
The `_isSequelizeMethod` check does not work for me. Even if it did, I still think we should "cloneDeep" regardless if a function was given or not. This also protects users from passing a function and returning the same object reference.

Suggested implementation:

``` js
      defaults = _.mapValues(this.Model._defaultValues, function(fn) {
        return _.cloneDeep(fn());
      });
```

rplopes replied:
I tried `_.clone` and `_.cloneDeep` before, but that would cause an issue with sequelize methods, making an integration test fail and potentially introducing regressions. More info in https://github.com/seegno-forks/sequelize/pull/3#issuecomment-166697719 and https://github.com/seegno-forks/sequelize/pull/3#issuecomment-167829881

nunofgs replied:
Can you investigate further please? I was able to run all unit and integration tests.

rplopes replied:
Using that code, I get an integration test for SQLite failing, when running `npm t`:

```
  1) [SQLITE] Model create is possible to use functions as default values:
     SequelizeValidationError: string violation: year cannot be an array or an object
      at lib/instance-validator.js:72:14
  From previous event:
      at Promise.then (lib/promise.js:21:17)
      at Context.<anonymous> (test/integration/model/create.test.js:811:53)
```

It's the same error I was getting before. From my previous research, it seems to be because `valueFn()._isSequelizeMethod` isn't being correctly cloned. If this is passing on your machine, it might be a bug with some software version on mine.

nunofgs replied:
Nevermind, it was a previous edit of mine. LGTM.
